### PR TITLE
Validating over gRPC

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Abstracts the Validator Client away from Validator Coordinator and allows each Validator Client to be responsible for its own private key. Validator Clients communicate with Coordinator over gRPC.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,9 +1,9 @@
 [submodule "eth-reference-tests/src/test/resources/eth2.0-tests"]
 	path = eth-reference-tests/src/test/resources/eth2.0-tests
 	url = https://github.com/ethereum/eth2.0-tests.git
-[submodule "ganache-cli"]
-	path = ganache-cli
-	url = https://github.com/trufflesuite/ganache-cli.git
 [submodule "bacon-vision"]
 	path = bacon-vision
 	url = https://github.com/baconviewer/bacon-vision.git
+[submodule "ganache-cli"]
```
Comment (cemozerr) [inline .gitmodules:7]: I had to do this due to the Jenkins error, but this somehow fixed it.

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/AttestationUtil.java [hunk 2]
```diff
@@ -44,106 +45,102 @@ public static boolean verifyAttestation(BeaconState state, Attestation attestati
...
+      epochBoundaryRoot = block.signed_root("signature");
+    } else {
+      epochBoundaryRoot = BeaconStateUtil.get_block_root(state, epochStartSlot);
+    }
+    UnsignedLong sourceEpoch = state.getCurrent_justified_epoch();
+    Bytes32 sourceRoot = state.getCurrent_justified_root();
+
+    // Set attestation data
+    // TODO: change the generic shard number and crosslink usage to something safer
+    AttestationData attestationData =
```
Comment (jrhea) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/AttestationUtil.java:94]: Make an issue for this TODO when u get a sec

File: validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java [hunk 3]
```diff
@@ -74,12 +88,13 @@ public ValidatorCoordinator(ServiceConfig config, ChainStorageClient store) {
...
 
-    initializeValidators();
-
     stateTransition = new StateTransition(printEnabled);
+
+    initializeValidators();
   }
 
   @Subscribe
+  // TODO: make sure blocks that are produced right even after new slot to be pushed.
```
Comment (jrhea) [inline validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java:96]: I don't understand this todo, but i respect it

File: validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java [hunk 4]
```diff
@@ -101,150 +116,173 @@ public void onNewHeadStateEvent(HeadStateEvent headStateEvent) {
...
-        AttestationUtil.createAttestations(headState, headBlock, validatorSet);
+    List<Triple<BLSPublicKey, Integer, CrosslinkCommittee>> attesters =
+        AttestationUtil.getAttesterInformation(headState, validatorSet);
+    AttestationData genericAttestationData =
+        AttestationUtil.getGenericAttestationData(headState, headBlock);
 
-    for (Attestation attestation : attestations) {
-      this.eventBus.post(attestation);
-    }
+    CompletableFuture.runAsync(
```
Comment (jrhea) [inline validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java:123]: welcome to stage 2

File: validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java [hunk 5]
```diff
@@ -101,150 +116,173 @@ public void onNewHeadStateEvent(HeadStateEvent headStateEvent) {
...
         startIndex
             + (numValidators / numNodes - 1)
             + toIntExact(Math.round((double) nodeCounter / Math.max(1, numNodes - 1)));
     endIndex = Math.min(endIndex, numValidators - 1);
-    // int startIndex = 0;
-    // int endIndex = numValidators-1;
     LOG.log(Level.DEBUG, "startIndex: " + startIndex + " endIndex: " + endIndex);
     for (int i = startIndex; i <= endIndex; i++) {
       BLSKeyPair keypair = BLSKeyPair.random(i);
+      int port = 50000 + i;
```
Comment (jrhea) [inline validator/coordinator/src/main/java/tech/pegasys/artemis/validator/coordinator/ValidatorCoordinator.java]: make a todo to externalize port 5000 as the starting point.  or make it a constant

**Review by jrhea — APPROVED** [review]
made a couple of comments, but lgtm

**Pull request merged**